### PR TITLE
Upgrade Android Components

### PR DESCRIPTION
**Pull request by Holidaybear**
- Upgrade Android Components to 0.51.0
- Add `lib-fetch-httpurlconnection` because `HttpURLConnectionTelemetryClient` was removed
- Modify UrlInputFragment because `DomainAutoCompleteProvider` was deprecated & OnFilterListener 2nd param was removed
- Modify UrlInputFragment layout in order to pass UI test (SearchFieldTest)

### CODE REVIEW

**Comment by Holidaybear** [comment]
related to #3137

**Comment by cnevinc** [comment]
lgtm

File: app/src/main/java/org/mozilla/focus/urlinput/UrlInputFragment.kt [hunk 1]
```diff
@@ -188,11 +188,14 @@ class UrlInputFragment : Fragment(), UrlInputContract.View, View.OnClickListener
...
-            when {
-                it.isEmpty() || !URLUtil.isValidUrl(urlView.text.toString()) -> urlView.text.toString()
-                else -> it
+        val input = urlView.autocompleteResult?.let { result ->
+            if (result.text.isEmpty() || !URLUtil.isValidUrl(urlView.text.toString())) {
+                urlView.text.toString()
+            } else {
+                result.text
             }
+        } ?: run {
```
Comment (pharspher) [inline app/src/main/java/org/mozilla/focus/urlinput/UrlInputFragment.kt]: run() seems unnecessary here?

File: components/feature/tabs/src/main/java/org/mozilla/rocket/tabs/TabViewEngineObserver.kt [hunk 2]
```diff
@@ -83,7 +83,7 @@ class TabViewEngineObserver(
 
     override fun onExternalResource(
         url: String,
-        fileName: String?,
+        fileName: String,
```
Comment (pharspher) [inline components/feature/tabs/src/main/java/org/mozilla/rocket/tabs/TabViewEngineObserver.kt]: it seems the caller of this method get all parameters from a Download object, which declares its name field as nullable, is it safe here to change it to non-optional type?

File: app/src/main/java/org/mozilla/focus/urlinput/UrlInputFragment.kt (hunk 1 above)
Comment (Holidaybear) [reply app/src/main/java/org/mozilla/focus/urlinput/UrlInputFragment.kt]: If `urlView.autocompleteResult == null`, will run this statement.

File: components/feature/tabs/src/main/java/org/mozilla/rocket/tabs/TabViewEngineObserver.kt (hunk 2 above)
Comment (Holidaybear) [reply components/feature/tabs/src/main/java/org/mozilla/rocket/tabs/TabViewEngineObserver.kt]: After upgrading android-components, they changed its `fileName` field to non-null from `EngineSession.Observer`

**Pull request merged**